### PR TITLE
chore: fix publishing workflows

### PR DESCRIPTION
**Pull request by jonybur**
# Description

Some publishing workflows are currently broken. This PR fixes them. It bumps the rust-toolchain version to 1.67.0 and makes workflows more similar between each other. Also refactors some bits and installs proper dependencies.

## Problem\*

Fixes publishing for following packages:
- [x] abi_wasm
- [x] acvm-js
- [x] noir-js
- [x] noir-wasm
- [x] source-resolver

## Documentation

- [ ] This PR requires documentation updates when merged.

  

  - [ ] I will submit a noir-lang/docs PR.

  

  - [ ] I will request for and support Dev Rel's help in documenting this PR.

  
  

## Additional Context

# PR Checklist\*

- [ ] I have tested the changes locally.
- [ ] I have formatted the changes with [Prettier](https://prettier.io/) and/or `cargo fmt` on default settings.

### CODE REVIEW

File: .github/workflows/publish-noir-js.yml [hunk 1]
```diff
@@ -17,9 +17,13 @@ jobs:
       - name: Install jq
         run: sudo apt-get install jq
 
+      - name: Setup toolchain
+        uses: dtolnay/rust-toolchain@master
+        with:
+          toolchain: 1.67.0
```
Comment (TomAFrench) [inline .github/workflows/publish-noir-js.yml:23]: Why the bump to 1.67.0?

File: .github/workflows/publish-noir-js.yml [hunk 2]
```diff
@@ -17,9 +17,13 @@ jobs:
...
+
       - name: Install cargo and wasm-bindgen
         run: |
-          curl --proto '=https' --tlsv1.2 -sSf https://sh.rustup.rs | sh -s -- -y
           source $HOME/.cargo/env
           cargo install -f wasm-bindgen-cli --version 0.2.86
```
Comment (TomAFrench) [inline .github/workflows/publish-noir-js.yml:28]:
```
      - name: Install wasm-bindgen-cli
        uses: taiki-e/install-action@v2
        with:
          tool: wasm-bindgen-cli@0.2.86
```

File: .github/workflows/publish-noir-wasm.yml [hunk 3]
```diff
@@ -21,23 +21,27 @@ jobs:
             target/
           key: ${{ runner.os }}-cargo-${{ hashFiles('**/Cargo.lock') }}
 
+      - name: Authenticate with npm
+        run: echo "//registry.npmjs.org/:_authToken=${{ secrets.NPM_TOKEN }}" > ~/.npmrc
+
```
Comment (TomAFrench) [inline .github/workflows/publish-noir-wasm.yml:26]: Let's move this closer to where it's used.

File: .github/workflows/publish-source-resolver.yml [hunk 4]
```diff
@@ -11,6 +11,12 @@ jobs:
       - name: Checkout
         uses: actions/checkout@v4
 
+      - name: Setup Node.js
+        uses: actions/setup-node@v3
+        with:
+          registry-url: "https://registry.npmjs.org"
+          node-version: 18.15
```
Comment (TomAFrench) [inline .github/workflows/publish-source-resolver.yml:18]: Why is this necessary here but not on the other workflows?

File: .github/workflows/publish-noir-js.yml (hunk 2 above)
Comment (TomAFrench) [reply .github/workflows/publish-noir-js.yml:28]: This removes the need for bumping the MSRV.

File: .github/workflows/publish-noir-js.yml (hunk 1 above)
Comment (jonybur) [reply .github/workflows/publish-noir-js.yml:23]: rustc would throw an error when installed on 1.66.1:

https://github.com/noir-lang/noir/actions/runs/6375729739/job/17301824921

File: .github/workflows/publish-noir-js.yml (hunk 2 above)
Comment (jonybur) [reply .github/workflows/publish-noir-js.yml]: Revised

**Comment by kobyhallx** [comment]
Would this workflow not benefit from same build mechanism as used for js tests?

File: .github/workflows/publish-noir-js.yml [hunk 5]
```diff
@@ -17,11 +17,15 @@ jobs:
...
 
-      - name: Install cargo and wasm-bindgen
-        run: |
-          curl --proto '=https' --tlsv1.2 -sSf https://sh.rustup.rs | sh -s -- -y
-          source $HOME/.cargo/env
-          cargo install -f wasm-bindgen-cli --version 0.2.86
+      - name: Setup toolchain
+        uses: dtolnay/rust-toolchain@master
+        with:
+          toolchain: 1.67.0
```
Comment (TomAFrench) [inline .github/workflows/publish-noir-js.yml:23]:
```suggestion
        uses: dtolnay/rust-toolchain@1.66.0
```

We don't want to increase the MSRV.

File: .github/workflows/publish-noir-js.yml [hunk 6]
```diff
@@ -34,6 +38,9 @@ jobs:
       - name: Build noirc_abi
         run: yarn workspace @noir-lang/noirc_abi build
 
+      - name: Build acvm_js
+        run: yarn workspace @noir-lang/acvm_js build
+
       - name: Build noir_js
         run: yarn workspace @noir-lang/noir_js build
```
Comment (TomAFrench) [inline .github/workflows/publish-noir-js.yml:45]: Noir_js needs the types and backend packages in order to be built. Are you sure that this works?

File: .github/workflows/publish-source-resolver.yml (hunk 4 above)
Comment (TomAFrench) [reply .github/workflows/publish-source-resolver.yml:18]: Can you open an issue to standardise this?

File: .github/workflows/publish-noir-js.yml (hunk 6 above)
Comment (jonybur) [reply .github/workflows/publish-noir-js.yml:43]: This seems to be working as per this dry run: https://github.com/noir-lang/noir/actions/runs/6376154044/job/17302810816

Comment (kevaundray) [reply .github/workflows/publish-noir-js.yml:43]: This can sometimes happen due to caching

File: .github/workflows/publish-abi_wasm.yml [hunk 7]
```diff
@@ -42,10 +41,11 @@ jobs:
         run: |
           mkdir temp_publish_dir
           cp -r ${{ env.BUILD_OUTPUT_PATH }}/* temp_publish_dir/
+          touch temp_publish_dir/yarn.lock
```
Comment (kevaundray) [inline .github/workflows/publish-abi_wasm.yml:44]: Why is this needed?

Comment (jonybur) [reply .github/workflows/publish-abi_wasm.yml:44]: `yarn npm publish` will throw an error otherwise, as temp_publish_dir is not part of the workspaces.
The accepted solution for this is to create an empty yarn.lock so that yarn understands that the project is separate from the monorepo.

Comment (kobyhallx) [reply .github/workflows/publish-abi_wasm.yml:44]: An alternative would be to copy those files back to originating folders and publish from there?
Similar to this:
[how integration test builds packages](https://github.com/noir-lang/noir/blob/8640bc77380b0abea888ad8e7c3dc6bb2f0d7c7c/.github/workflows/test-integration.yml#L136)

File: .github/workflows/publish-noir-js.yml [hunk 8]
```diff
@@ -34,6 +36,9 @@ jobs:
       - name: Build noirc_abi
         run: yarn workspace @noir-lang/noirc_abi build
 
+      - name: Build acvm_js
```
Comment (kobyhallx) [inline .github/workflows/publish-noir-js.yml:39]: We are building [acvm_js](https://github.com/noir-lang/noir/blob/8640bc77380b0abea888ad8e7c3dc6bb2f0d7c7c/.github/workflows/publish-acvm-js.yml#L36) one way when publishing and another way here. Why not unify approach and benefit from cache if done in sequence.

**Comment by kobyhallx** [comment]
I think we should close this item as is (assuming functioning correctly). We would benefit from refactoring set of files into single workflow where we build, cache, test and publish (depending on occasion).

File: .github/workflows/publish-source-resolver.yml (hunk 4 above)
Comment (jonybur) [reply .github/workflows/publish-source-resolver.yml:18]: https://github.com/orgs/noir-lang/projects/1/views/20?sliceBy%5Bvalue%5D=Week+40&pane=issue&itemId=40475271

File: .github/workflows/publish-abi_wasm.yml (hunk 7 above)
Comment (jonybur) [reply .github/workflows/publish-abi_wasm.yml:44]: I think that might fail due to the nature of nix's /result folder. In your example we are not generating a .tar inside of nix's result folder.

**Review by kobyhallx — APPROVED** [review]
